amida-tech/amida-notification-microservice
# WIP: Feature/ser 270 postgres to mongo

If this PR fixes a bug, you _must_ add test cases representative of the bug.
### What's this PR do?
#### So far
- brings in `mongoose` and initiates connection on server start
- defines `user` mongoose model that includes data from both `users` and `devices` postges tables
- ~does not replace unused `notifications` table~ Turns out it was used! (Oops!) This is also replaced now.
- removes unused `users/sendPushNotification` endpoint
- replaces functionality of existing controllers
- uses async/await syntax in controllers (with wrapper function to catch errors and pass them to `next`)
- passes existing tests
#### Questions
- I commented out the bluebird promise override because it didn't seem like we needed it anymore. Thoughts?
  - EDIT: Aaron agreed we don't need it - I will remove it.
- Do we want to keep the old postgres migrations around? probably right? so I guess I should move `sequelize` to be a `devDependency`?
  - EDIT:
  - I moved everything sequelize related into its own directory, and moved all sql dependencies to devDependencies
  - I left the PG Envvars in config.js with a comment to delete them when we don't need them anymore
- Take a look at the pattern I used for connecting to mongoose, registering models, and using those models in the controllers. It was loosely modeled after how it's done in orange-api, but I would appreciate input.
  - EDIT: I refactored it to be more friendly IMO - would still appreciate feedback
- Should the service log an error and quit if we can't connect to database? probably right?
  - EDIT: Aaron agreed, I will do this.
#### Remaining tasks
- [x] Add more tests to verify currently untested functionality
- [x] Create migration script from postgres to mongo
- [x] Implement pattern for future mongo migrations
- [x] Verify connection to mongo is encrypted when specified
  - EDIT: unnecessary - MongoClient refuses to connect if `ssl = true` and the mongodb instance doesn't support it
- [x] Support separate User/Pass/DbName envvars
- [x] Recreate `create_missing_users` script
- [x] Edit kubernetes (and other?) deployment configs/scripts
- [ ] probably more...?
### Related JIRA tickets:
- https://jira.amida.com/browse/SER-270
### How should this be manually tested?
- Run all the sql and mongo scripts and verify they work as expected
- Run the migration on our dev orange deployment and test the features of orange around notifications would probably be a good way of testing parity.
### Any background context you want to provide?
- https://jira.amida.com/browse/SER-268
